### PR TITLE
Feature/dope 212 parameter extraction

### PR DESCRIPTION
**Pull request by jansigi**

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/PrefixOperator.kt [hunk 1]
```diff
@@ -1,8 +1,15 @@
 package ch.ergon.dope.resolvable.operator
 
+import ch.ergon.dope.DopeQuery
 import ch.ergon.dope.resolvable.Resolvable
 import ch.ergon.dope.resolvable.formatToQueryStringWithBrackets
 
 open class PrefixOperator(private val symbol: String, private val argument: Resolvable) {
-    fun toPrefixQueryString() = formatToQueryStringWithBrackets(symbol, argument)
+    fun toPrefixQueryString(): DopeQuery {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/operator/PrefixOperator.kt]: rename to `toPrefixDopeQuery`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/InfixOperator.kt [hunk 2]
```diff
@@ -1,13 +1,21 @@
...
+import ch.ergon.dope.resolvable.formatMinimumTwoToQueryString
 import ch.ergon.dope.validtype.ValidType
 
 open class InfixOperator(
     private val left: TypeExpression<out ValidType>,
     private val symbol: String,
     private val right: TypeExpression<out ValidType>,
 ) {
-    fun toInfixQueryString() = formatToQueryString(left.toQueryString(), symbol, right.toQueryString())
+    fun toInfixQuery(): DopeQuery {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/operator/InfixOperator.kt]: rename to `toInfixDopeQuery`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/Concat2Expression.kt [hunk 3]
```diff
@@ -11,7 +12,22 @@ class Concat2Expression(
...
+        val stringDopeQuery = string.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(
+                symbol = "CONCAT2",
+                separatorDopeQuery,
+                stringDopeQuery,
+                *stringsDopeQuery.toTypedArray(),
+            ),
+            parameters = separatorDopeQuery.parameters + stringDopeQuery.parameters + stringsDopeQuery.fold(emptyMap()) { map, field ->
+                map + field.parameters
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/Concat2Expression.kt:27]: same comments as in the GroupByClause

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/ConcatExpression.kt [hunk 4]
```diff
@@ -10,7 +11,22 @@ class ConcatExpression(
...
+        return DopeQuery(
+            queryString = toFunctionQueryString(
+                symbol = "CONCAT",
+                firstStringDopeQuery,
+                secondStringDopeQuery,
+                *stringTypesDopeQuery.toTypedArray(),
+            ),
+            parameters = firstStringDopeQuery.parameters + secondStringDopeQuery.parameters + stringTypesDopeQuery.fold(
+                emptyMap(),
+            ) { map, field -> map + field.parameters },
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/ConcatExpression.kt:27]: rename `map`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt [hunk 5]
```diff
@@ -12,7 +13,15 @@ class LpadExpression(
...
     private val prefix: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "LPAD", inStr, size, extra = prefix)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val sizeDopeQuery = size.toQuery()
+        val prefixDopeQuery = prefix?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "LPAD", inStrDopeQuery, sizeDopeQuery, extra = prefixDopeQuery),
+            parameters = inStrDopeQuery.parameters + sizeDopeQuery.parameters + (prefixDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt:22]: there is an extension-function, `orEmpty()`: `...prefixDopeQuery?.parameters.orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LtrimExpression.kt [hunk 6]
```diff
@@ -9,7 +10,14 @@ class LtrimExpression(
     private val inStr: TypeExpression<StringType>,
     private val extra: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "LTRIM", inStr, extra = extra)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val extraDopeQuery = extra?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "LTRIM", inStrDopeQuery, extra = extraDopeQuery),
+            parameters = inStrDopeQuery.parameters + (extraDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LtrimExpression.kt:18]: `orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt [hunk 7]
```diff
@@ -8,7 +9,13 @@ import ch.ergon.dope.validtype.StringType
 class NowStrExpression(
     private val format: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString("NOW_STR", extra = format)
+    override fun toQuery(): DopeQuery {
+        val formatDopeQuery = format?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString("NOW_STR", extra = formatDopeQuery),
+            parameters = formatDopeQuery?.parameters ?: emptyMap(),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:16]: `orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RpadExpression.kt [hunk 8]
```diff
@@ -12,7 +13,15 @@ class RpadExpression(
...
     private val extra: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "RPAD", inStr, size, extra = extra)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val sizeDopeQuery = size.toQuery()
+        val extraDopeQuery = extra?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "RPAD", inStrDopeQuery, sizeDopeQuery, extra = extraDopeQuery),
+            parameters = inStrDopeQuery.parameters + sizeDopeQuery.parameters + (extraDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RpadExpression.kt:22]: `orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RtrimExpression.kt [hunk 9]
```diff
@@ -9,7 +10,14 @@ class RtrimExpression(
     private val inStr: TypeExpression<StringType>,
     private val extra: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "RTRIM", inStr, extra = extra)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val extraDopeQuery = extra?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "RTRIM", inStrDopeQuery, extra = extraDopeQuery),
+            parameters = inStrDopeQuery.parameters + (extraDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RtrimExpression.kt:18]: `orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/SplitExpression.kt [hunk 10]
```diff
@@ -9,7 +10,14 @@ class SplitExpression(
     private val inStr: TypeExpression<StringType>,
     private val inSubstring: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "SPLIT", inStr, extra = inSubstring)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val inSubstringDopeQuery = inSubstring?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "SPLIT", inStrDopeQuery, extra = inSubstringDopeQuery),
+            parameters = inStrDopeQuery.parameters + (inSubstringDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/SplitExpression.kt:18]: `orEmpty()`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/TrimExpression.kt [hunk 11]
```diff
@@ -10,7 +11,14 @@ class TrimExpression(
     private val inStr: TypeExpression<StringType>,
     private val char: TypeExpression<StringType>? = null,
 ) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = toFunctionQueryString(symbol = "TRIM", inStr, extra = char)
+    override fun toQuery(): DopeQuery {
+        val inStrDopeQuery = inStr.toQuery()
+        val charDopeQuery = char?.toQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol = "TRIM", inStrDopeQuery, extra = charDopeQuery),
+            parameters = inStrDopeQuery.parameters + (charDopeQuery?.parameters ?: emptyMap()),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/TrimExpression.kt:19]: `orEmpty()`

**Review by pgruntz — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/Resolvable.kt [hunk 12]
```diff
@@ -1,5 +1,7 @@
 package ch.ergon.dope.resolvable
 
+import ch.ergon.dope.DopeQuery
+
 interface Resolvable {
-    fun toQueryString(): String
+    fun toQuery(): DopeQuery
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/Resolvable.kt:6]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt [hunk 13]
```diff
@@ -1,15 +1,31 @@
...
+        val fieldDopeQuery = field.toQuery()
+        val fieldsDopeQuery = fields.map { it.toQuery() }
+        return DopeQuery(
+            queryString = formatMinimumTwoToQueryString(
+                parentDopeQuery.queryString,
+                "GROUP BY",
+                fieldDopeQuery.queryString,
+                *fieldsDopeQuery.map { it.queryString }.toTypedArray(),
+            ),
+            parameters = fieldDopeQuery.parameters + fieldsDopeQuery.fold(emptyMap()) { map, field -> map + field.parameters } +
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt:27]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt [hunk 14]
```diff
@@ -1,17 +1,48 @@
...
+    override fun toQuery(): DopeQuery {
+        val expressionDopeQuery = expression.toQuery()
+        val expressionsDopeQuery = expressions.map { it.toQuery() }
+        return DopeQuery(
+            queryString = formatToQueryString(
+                "SELECT",
+                expressionDopeQuery.queryString,
+                *expressionsDopeQuery.map { it.queryString }.toTypedArray(),
+            ),
+            parameters = expressionDopeQuery.parameters + expressionsDopeQuery.fold(emptyMap()) { map, field -> map + field.parameters },
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:19]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt [hunk 15]
```diff
@@ -1,17 +1,48 @@
...
+    override fun toQuery(): DopeQuery {
+        val expressionsDopeQuery = expressions.map { it.toQuery() }
+        val expressionDopeQuery = expression.toQuery()
+        return DopeQuery(
+            queryString = formatToQueryString(
+                "SELECT DISTINCT",
+                expressionDopeQuery.queryString,
+                *expressionsDopeQuery.map { it.queryString }.toTypedArray(),
+            ),
+            parameters = expressionDopeQuery.parameters + expressionsDopeQuery.fold(emptyMap()) { map, field -> map + field.parameters },
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/SelectClause.kt:45]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt [hunk 16]
```diff
@@ -1,28 +1,71 @@
...
+    constructor(collection: Collection<Any>, parameterName: String?) {
+        this.value = collection
+        this.parameterName = parameterName
+    }
 
-fun resetCounter() {
-    ParameterCounter.count = 1
+    override fun toQuery(): DopeQuery = when (parameterName) {
+        null -> {
+            val count = "\$${ParameterManager.count}"
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:46]: i renamed it to `unnamedParameterCount`

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/PrefixOperator.kt (hunk 1 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/operator/PrefixOperator.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/InfixOperator.kt (hunk 2 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/operator/InfixOperator.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/ConcatExpression.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/ConcatExpression.kt:27]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt:27]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/ReturningClause.kt [hunk 17]
```diff
@@ -1,15 +1,31 @@
...
+        val fieldDopeQuery = field.toQuery()
+        val parentDopeQuery = parentClause.toQuery()
+        return DopeQuery(
+            queryString = formatMinimumTwoToQueryString(
+                parentDopeQuery.queryString,
+                "RETURNING",
+                fieldDopeQuery.queryString,
+                *fieldsDopeQuery.map { it.queryString }.toTypedArray(),
+            ),
+            parameters = fieldDopeQuery.parameters + fieldsDopeQuery.fold(emptyMap()) { map, field -> map + field.parameters } +
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/ReturningClause.kt:27]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/Concat2Expression.kt (hunk 3 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/Concat2Expression.kt:27]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt (hunk 5 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt:22]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LtrimExpression.kt (hunk 6 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LtrimExpression.kt:18]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt (hunk 7 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:16]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RpadExpression.kt (hunk 8 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RpadExpression.kt:22]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RtrimExpression.kt (hunk 9 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RtrimExpression.kt:18]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/SplitExpression.kt (hunk 10 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/SplitExpression.kt:18]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/TrimExpression.kt (hunk 11 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/TrimExpression.kt:19]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/CountExpression.kt [hunk 18]
```diff
@@ -8,11 +9,17 @@ import ch.ergon.dope.validtype.ValidType
...
+        return DopeQuery(
+            queryString = formatToQueryStringWithBrackets("COUNT", fieldDopeQuery.queryString),
+            parameters = fieldDopeQuery.parameters,
+        )
+    }
 }
 
 class CountAsteriskExpression : AggregateExpression {
-    override fun toQueryString(): String = "COUNT($ASTERISK_STRING)"
+    override fun toQuery(): DopeQuery = DopeQuery("COUNT($ASTERISK_STRING)", emptyMap())
```
[the remark this answers is not in the document]
Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/CountExpression.kt]: `DopeQuery(queryString = "COUNT($ASTERISK_STRING)", parameters = emptyMap())`

[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/CountExpression.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/AsteriskExpression.kt [hunk 19]
```diff
@@ -20,5 +21,5 @@ class AsteriskExpression : Expression {
         queryString = ASTERISK_STRING
     }
 
-    override fun toQueryString(): String = queryString
+    override fun toQuery(): DopeQuery = DopeQuery(queryString, emptyMap())
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/AsteriskExpression.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt [hunk 20]
```diff
@@ -11,7 +12,7 @@ import ch.ergon.dope.validtype.ValidType
 
 class Primitive<T : ValidType> : TypeExpression<T> {
     private val queryString: String
-    override fun toQueryString(): String = queryString
+    override fun toQuery(): DopeQuery = DopeQuery(queryString, emptyMap())
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/QueryStringBuilder.kt [hunk 21]
```diff
@@ -1,16 +1,14 @@
...
-fun formatToQueryString(left: Resolvable, right: String) = "${left.toQueryString()} $right"
-fun formatToQueryString(left: String, symbol: String, right: String) = "$left $symbol $right"
-fun formatToQueryString(left: Resolvable, symbol: String, vararg right: Resolvable) =
-    "${left.toQueryString()} $symbol ${right.joinToString { it.toQueryString() }}"
-fun formatToQueryString(symbol: String, vararg argument: Resolvable) =
-    "$symbol ${argument.joinToString(separator = ", ") { it.toQueryString() }}"
+fun formatToQueryString(left: String, vararg right: String) =
+    "$left ${right.joinToString()}"
+
+fun formatMinimumTwoToQueryString(left: String, symbol: String, vararg right: String) =
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/QueryStringBuilder.kt]: done

**Pull request merged**